src/s2/s2polygon.cc

```diff
@@ -1282,41 +1291,48 @@ unique_ptr<S2Polygon> S2Polygon::DestructiveApproxUnion(
 }
 
 unique_ptr<S2Polygon> S2Polygon::DestructiveUnion(
-    vector<std::unique_ptr<S2Polygon>> polygons,
+    vector<unique_ptr<S2Polygon>> polygons,
     const S2Builder::SnapFunction& snap_function) {
-  // Effectively create a priority queue of polygons in order of number of
-  // vertices.  Repeatedly union the two smallest polygons and add the result
-  // to the queue until we have a single polygon to return.
-  using QueueType = std::multimap<int, unique_ptr<S2Polygon>>;
-  QueueType queue;  // Map from # of vertices to polygon.
-  for (auto& polygon : polygons)
-    queue.insert(std::make_pair(polygon->num_vertices(), std::move(polygon)));
+  if (polygons.empty())
+    return make_unique<S2Polygon>();
+
+  // Create a priority queue of polygons in order of number of vertices.
+  // Repeatedly union the two smallest polygons and add the result to the
+  // queue until we have a single polygon to return.  We use the index
+  // into `polygons` as the second component of the pair instead of a
+  // pointer so the comparison is deterministic rather than depending
+  // on the polygon's address.  (Some clients depend on determinism.)
+  using Pair = pair<int, size_t>;  // (num vertices, `polygons` index)
+  using PriorityQueue = priority_queue<Pair, vector<Pair>, greater<Pair>>;
+  PriorityQueue queue;  // Least vertices is `top`.
+  for (size_t i = 0; i < polygons.size(); ++i) {
+    queue.emplace(polygons[i]->num_vertices(), i);
+  }
 
   while (queue.size() > 1) {
     // Pop two simplest polygons from queue.
-    QueueType::iterator smallest_it = queue.begin();
-    int a_size = smallest_it->first;
-    unique_ptr<S2Polygon> a_polygon(std::move(smallest_it->second));
-    queue.erase(smallest_it);
-    smallest_it = queue.begin();
-    int b_size = smallest_it->first;
-    unique_ptr<S2Polygon> b_polygon(std::move(smallest_it->second));
-    queue.erase(smallest_it);
+    const Pair a = queue.top();
+    queue.pop();
+    const Pair b = queue.top();
+    queue.pop();
 
     // Union and add result back to queue.
     auto union_polygon = make_unique<S2Polygon>();
-    union_polygon->InitToUnion(*a_polygon, *b_polygon, snap_function);
-    queue.insert(std::make_pair(a_size + b_size, std::move(union_polygon)));
+    union_polygon->InitToUnion(*polygons[a.second], *polygons[b.second],
+                               snap_function);
+
+    // Replace `a` with the union, and clear `b` to reclaim memory.
+    polygons[a.second] = std::move(union_polygon);
+    polygons[b.second].reset();
+
     // We assume that the number of vertices in the union polygon is the
     // sum of the number of vertices in the original polygons, which is not
     // always true, but will almost always be a decent approximation, and
     // faster than recomputing.
+    queue.emplace(a.first + b.first, a.second);
   }
```

Comment:
I see it old code, but is num_vertices slow for polygon? I thought it's computed in ctor

Reply:
Good catch. It looks like `num_vertices` changed back in 2009, so this is out of date.

Reply:
This will be fixed separately.